Convert `ServerFormsDetailsFetcher` to Kotlin

seadowg opened the pull request:
~~Blocked by #4906~~

We've seen a couple of recent crashes that we'd probably have caught with Kotlin's null safety/strictness, so I wanted to make sure we got this code out of Java. I also restructured the tests a little as they've grown in size and were feeling a little awkward.

#### What has been done to verify that this works as intended?

Ran tests.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

We've done a lot of QA in this area recently, and it seems like our test coverage is pretty well lined up with our regression checks as we've mostly been catching things in production once they hit users using custom servers. I think it'd be good for the reviewer to think of cases we might not have covered. This would be a good opportunity to add tests at the unit level or in `BadServerTest`.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Convert class to Kotlin

seadowg committed:
Reformat code for Kotlin

seadowg committed:
Remove nullables and vars

seadowg committed:
Convert test to Kotlin

seadowg committed:
Pull out individual fixtures to focus tests

seadowg committed:
Convert FormListItem to data class

seadowg committed:
Use copy to cut down number of fixtures

seadowg committed:
Fix mock usage

seadowg commented on forms/src/main/java/org/odk/collect/forms/FormListItem.kt:
```diff
@@ -0,0 +1,10 @@
+    val downloadURL: String,
+    val formID: String,
+    val version: String,
+    val hashWithPrefix: String?,
```
It'd be good to think about if any of the other fields (here and in the other new data classes) should `String?` here. If so, we missing test coverage of whatever case that would be from.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.kt:
```diff
@@ -0,0 +1,146 @@
+            val forms = formsRepository.getAllNotDeletedByFormId(listItem.formID)
+            val thisFormAlreadyDownloaded = forms.isNotEmpty()
+            val isNewerFormVersionAvailable = listItem.hashWithPrefix.let {
+                if (it == null || !it.startsWith("md5:")) {
```
I would move checking for that prefix `md5:` to [OpenRosaResponseParserImpl](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.java#L106) so that we deal with it only there. If that prefix does not exist we should return null like in case of empty value. What do you think?

seadowg replied:
That's an interesting idea. I'll have a play around with it.

grzesiek2010 replied:
I would even go the extra mile and remove that prefix there so that in `FormListItem` we have it without `md5:` and we don't need to call `substring` every time we use the hash.

grzesiek2010 replied:
...or pass it to `FormListItem` with the prefix but have a custom getter which would ignore it.

seadowg replied:
Yeah I was also thinking about a new type (like a `Hash`) that enforces the prefix but means we don't have to deal with it everywhere. I'll play around with a few of these ideas, see what I like, and then we can review!

grzesiek2010 replied:
If you decide to implement changes in `OpenRosaResponseParserImpl` can convert it to Kotlin as well,

seadowg replied:
Ok done!

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.kt:
```diff
@@ -0,0 +1,146 @@
+
+        val formList = formSource.fetchFormList()
+        return formList.map { listItem ->
+            val manifestFile = listItem.manifestURL.let {
```
I think you can simplify this:
```kotlin
val manifestFile = listItem.manifestURL?.let {
    getManifestFile(formSource, it)
}
```

seadowg replied:
Yes! I remember writing this and being like "I swear there's a cleaner way to do this I'm forgetting". `?.let` is a great trick - basically `map` from a `Maybe` monad.

seadowg committed:
Use cleaner way of mapping optional

seadowg committed:
Convert class to Kotlin

seadowg committed:
Return null hashes instead of invalid ones from parser

seadowg committed:
Introduce hash without prefix property

seadowg committed:
Introduce hash without prefix for ServerFormDetails

This changes the key we use for last update notification which means
that users might see a notification update they have seen before after
update. I don't think this feels like a big enough deal to work around.

seadowg committed:
Stop using hashWithPrefix from FormListItem

seadowg committed:
Remove hashWithPrefix

seadowg committed:
Make MediaFile has name explicit

seadowg committed:
Stop using MediaFile#hashWithPrefix

seadowg committed:
Remove MediaFile#hashWithPrefix

seadowg committed:
Add test to describe missing hash prefix for media files

seadowg commented on collect_app/src/androidTest/java/org/odk/collect/android/feature/smoke/BadServerTest.java:
```diff
@@ -42,4 +46,21 @@ public void whenHashNotIncludedInFormList_showError() {
+     (computed) hash with the server one assuming there is a prefix and end up comparing against
+     a substring.
+    */
+    public void whenMediaFileHasMissingPrefix_showsAsUpdated() {
```
This is probably not the behaviour we want, but it is the behaviour we have. I wrote this test against `master` to double-check. I'll pull out another issue, so we can discuss how to improve this.

seadowg replied:
https://github.com/getodk/collect/issues/4919

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+                        }
+                        "hash" -> {
+                            hash = XFormParser.getXMLText(child, true)
+                            if (hash != null && hash.length == 0) {
```
Don't you need to check if it starts with `md5:` here too?

grzesiek2010 replied:
Ah you want to improve it later https://github.com/getodk/collect/pull/4911/files#diff-97ea04c120d7632ab55eb3218d1ec61bf0c45056bc43f9ca6c6573feac4250cbR57?

seadowg replied:
Yeah exactly!

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+                    when (child.name) {
+                        "filename" -> {
+                            filename = XFormParser.getXMLText(child, true)
+                            if (filename != null && filename.length == 0) {
```
You can replace all those `length == 0` with `isEmpty()`

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+                        if (hash != null && (hash.length == 0 || !hash.startsWith("md5:"))) {
+                            hash = null
+                        } else {
+                            hash = hash.substring("md5:".length)
```
Since we use this in a few places maybe it's better to have a const for that `md5:`

seadowg replied:
Yeah I'd like to also have extracted an `Md5` data type, but I think I've gone deep enough. I'll do as you suggest and pull out the constant.

grzesiek2010 replied:
I'm not sure if putting that constant there is a good idea. It's like our spec that we add that prefix (`md5:`) and not `md5` or `md5-` for example right? So it should be a part of this package and a separate module (in the future) as we agreed on a few minutes later.

seadowg replied:
Ah I was thinking eventually we'd make `Md5` a class and have a `parsePrefixString` method on it or something. I think you're right that I'm jumping ahead there. Right now the prefix is only used in the parser so I'll move the constant there.

seadowg replied:
Ok done!

grzesiek2010 commented on collect_app/src/test/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImplTest.kt:
```diff
@@ -0,0 +1,39 @@
+import java.io.StringReader
+import java.lang.StringBuilder
+
+class OpenRosaResponseParserImplTest {
```
I think we should have more test here (it might be a separate issue). It looks like we haven't had `OpenRosaResponseParserImpl` covered with tests and you just added one for your case.

seadowg replied:
Yeah I think we pulled it out for testing the integration between it and `OpenRosaFormSource` so it never got its own tests (this was all originally untested code). This would be a nice object to backfill though, so I'll create an issue as you suggest.

seadowg replied:
https://github.com/getodk/collect/issues/4919

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+package org.odk.collect.android.openrosa
```
What do you think about moving it to its own module?

seadowg replied:
Definitely! I've looked at this a few times before, but I'm not sure if we've got all the ties to `collect_app` severed yet. I think I'd rather leave that for another time rather than adding to this PR.

grzesiek2010 replied:
Yes yes not in this pr.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+            }
+
+            val formListItem =
+                FormListItem(downloadUrl, formId, version!!, hash, formName, manifestUrl)
```
`version` might be null but it's not nullable in FormListItem.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/openrosa/OpenRosaResponseParserImpl.kt:
```diff
@@ -0,0 +1,220 @@
+                return null
+            }
+
+            val formListItem =
```
Does that make sense to construct this object ad all? You can just add it directly to the list:
`            formList.add(FormListItem(downloadUrl, formId, version!!, hash, formName, manifestUrl))
`

seadowg committed:
Replace length checks with isEmpty

seadowg committed:
Pull out constant for md5 prefix

seadowg committed:
Version should be nullable

seadowg committed:
Inline variable

seadowg committed:
Move prefix constant to where it's used
